### PR TITLE
Ability to use right alt aka. alt gr to reach all characters on non us/uk layouts

### PR DESCRIPTION
**Pull request by nuuttihenriksson**
Simply added possibility for non us/uk keyboards to reach characters.

Uses only libraries that were already included.

### CODE REVIEW

File: src/Keyboard.cpp [hunk 1]
```diff
@@ -232,14 +229,22 @@ size_t Keyboard_::press(uint8_t k)
...
 		}
-		if (k & 0x80) {						// it's a capital letter or other character reached with shift
-			_keyReport.modifiers |= 0x02;	// the left shift modifier
-			k &= 0x7F;
+		if (x & 0x80) {						// it's a capital letter or other character reached with shift
+			_keyReport.modifiers |= 0x02;			// the left shift modifier
+			x &= 0x7F;
+			k = (uint8_t)x;
+		} else if(x & 0x100) {					// it's a character reached with alt gr aka. right alt
+			_keyReport.modifiers |= 0x05;			// alt gr aka. right alt modifier
```
Comment (edgar-bonet) [inline src/Keyboard.cpp:243]: You mean

```c++
_keyReport.modifiers |= 0x40;
```

i.e.

```c++
_keyReport.modifiers |= (1<<(KEY_RIGHT_ALT-128));
```

**Review by edgar-bonet — COMMENTED** [review]
AltGr is already usable as

```c++
Keyboard.press(KEY_RIGHT_ALT);
```

How would this change make it easier? How would you use the new feature? It is not visible through the library API.

**Comment by potatoprogrammer1** [comment]
> AltGr is already usable as
> 
> ```c++
> Keyboard.press(KEY_RIGHT_ALT);
> ```
> 
> How would this change make it easier? How would you use the new feature? It is not visible through the library API.

I have added the alt gr as a modifier that can be applied to the ascii map. This means that instead of having to break up print commands because of characters such as /{}$€, they are managed through the ascii map in the .cpp file. 

This makes the library a lot more usable for people with keyboard layouts other than english.

As I said I implemented it in the Keyboard.cpp file. The asciimap as well as the memory map has been changed to 16bits to fit an extra check bit (as was done with shift modifier).

The shift modifier had the eith bit set as 1. Alt gr uses ninth bit set as 1. In addition to the 8bit -> 16bit change I also had to use memcpy_P to copy the number from the ascii map.

This change includes no new libraries and does not require a significant amount of resources.

**Comment by potatoprogrammer1** [comment]
> > I have added the alt gr as a modifier that can be applied to the ascii map.
> 
> But you have not modified the ASCII map, so it would seem the change serves no purpose. Or maybe you expect the users to modify the ASCII map themselves?
> 
> > characters such as /{}$€
> 
> `/`, `{`, `}` and `$` are ASCII characters, but `€` is not, thus it cannot fit in the ASCII map.
> 
> Furthermore, most computers nowadays use UTF-8 encoding for files, and “€” is then encoded as a sequence of three bytes: `0xe2`, `0x82` and `0xac`. It is not a “character” in the C++ sense (a `char`), meaning that
> 
> ```c++
> Keyboard.write('€');
> ```
> 
> is not valid C++, and
> 
> ```c++
> Keyboard.print("€");
> ```
> 
> would attempt to print three characters. In this case they happen to be all invalid, but printing `"ó"` for example (UTF-8: `0xc3`, `0xb3`) will be interpreted as `KEY_F2` followed by `KEY_TAB`.
> 
> Thus I ask again the same question: How would you use this new feature?

My bad. I was not suppose to include '€'. For example if I want to do print("C:\Users\Randomuser\Desktop\"); I would have to do 
Keyboard.print("C:");
Keyboard.press(KEY_RIGHT_ALT);
Keyboard.press(what ever the hex is for plus);
Keyboard.releaseAll();
Keyboard.print("Users");
...

When the alt gr / right alt is added into the code I can just print the line without any problems.

**Comment by nuuttihenriksson** [comment]
What is the status on this?

**Comment by per1234** [comment]
@nuuttihenriksson @edgar-bonet was this superseded by https://github.com/arduino-libraries/Keyboard/pull/53?

**Comment by nuuttihenriksson** [comment]
Outdated.

**Comment by nuuttihenriksson** [comment]
> @nuuttihenriksson @edgar-bonet was this superseded by #53?

Good points bt @edgar-bonet, was an early idea. #53 looks good.

**Comment by per1234** [comment]
Thanks for your speedy response, and for your contribution @nuuttihenriksson!

Regards, Per

**Pull request closed without merging**